content/blog/2020-11-30-super-optimization/index.md

```diff
@@ -0,0 +1,49 @@
+The paper we are discussing this week is ["Superoptimizer"][superopt] – A look at the Smallest Program” (Henry), this paper explains the purpose of superoptimizer, which converts an optimizing problem to a search space, how to reduce the search space and comparison with other common optimization techniques that are common at the publication of the paper. 
+
+
+## Super Optimization
+Super Optimization is finding the shortest program by exhaustive search inside a finite search space of the instruction set architecture (ISA).  Compare to other optimization techniques that focus on transformation and match, super optimization takes it to a whole different level.  Most optimizations techniques focus on the potential performance benefit on utilizing a transformation and ensuring correctness of the output by having proofs that the transformation does not affect the equivalence of the original program.  In modern compliers, this can ensure that the result is unchanged, but does not guarantee optimality of the output.  To address the need for the most optimal output, super optimization searches the entire sample spaces of the instruction set architecture for programs by generating all ombinations of eah instruction with various lengths, starting with the shortestlength.  
+
+Then super optimization then preform a test of the generated program determine if the generated program is equivalent with the orgnial program and if they are the same, then generated programwill be the optimal porgram for the ISA.  Although this soulds simple, verifing the generated program is a non trival task.  One of the sloution proposed in the paper was boolean test.  However as the complexity of some instructions have expressio of the order of 2^31 minterms (Massalin), it has caused a constriant on the number of instrcitios that can be generated and verified in a reasonable time to be limited to 3 instructions at the time of publication.  Therefore although the boolean test is intuitive, and correct a alternative test will be needed in order to test programs wiht larger amount of instructions.  
+
+The paper also has proposed a probailistict execution test in which the superoptimizer picks a test vectr and see if the output results are the same between the test and generated program.  Although this test is not compprehansive, the author claims that there was never a case where probailistict execution test passes while boolean test fails.  The time it takes for exhaustive search can be futher reduced by ulitization of pruing on the generated sequences that are equivalent to shorter sequences.  
```

Comment:
This paragraph mentions pruning but not how it works. How is it possible to prune some sequences—i.e., how does the superoptimizer know that the longer sequence is equivalent to some shorter sequence?